src/algebra/algebra/basic.lean

```diff
@@ -1404,20 +1404,49 @@ end linear_map
 end is_scalar_tower
 
 section restrict_scalars
-/- In this section, we describe restriction of scalars: if `S` is an algebra over `R`, then
-`S`-modules are also `R`-modules. -/
 
 section type_synonym
 variables (R A M : Type*)
 
-/--
-Warning: use this type synonym judiciously!
-The preferred way of working with an `A`-module `M` as `R`-module (where `A` is an `R`-algebra),
-is by `[module R M] [module A M] [is_scalar_tower R A M]`.
+/-- A type synonym representing the functor from the category of modules over `A`, regarded as a
+(semi)ring, to the category of modules over `A`, regarded as an `R`-algebra.
 
-When `M` is a module over a ring `A`, and `A` is an algebra over `R`, then `M` inherits a
-module structure over `R`, provided as a type synonym `module.restrict_scalars R A M := M`.
--/
+Warning: use this type synonym judiciously! Consider an example where we want to construct an
+`R`-linear map from `M` to `A`, given:
+```lean
+variables (R A M : Type*)
+variables [comm_semiring R] [semiring A] [algebra R A] [add_comm_monoid M] [module A M]
+```
+With the assumptions above we can't directly state our map as we have no `module R M` structure, but
+`restrict_scalars` permits it to be written as:
+```lean
+-- an `R`-module structure on `M` is provided by `restrict_scalars` which is compatible
+example : restrict_scalars R A M →ₗ[R] A := sorry
+```
+However, it is usually better just to add this extra structure as an argument:
+```lean
+-- an `R`-module structure on `M` and proof of its compatibility is provided by the user
+example [module R M] [is_scalar_tower R A M] : M →ₗ[R] A := sorry
+```
+The advantage of the second approach is that it defers the duty of providing the missing typeclasses
+`[module R M] [is_scalar_tower R A M]`. If some concrete `M` naturally carries these (as is often
+the case) then we have avoided `restrict_scalars` entirely. If not, we can pass
+`restrict_scalars R A M` later on instead of `M`.
+
+Note that this means we almost always want to state definitions and lemmas in the language of
+`is_scalar_tower` rather than `restrict_scalars`.
+
+Mathematically, when `A` is an `R`-algebra, there are two different functors:
+ 1. The functor from the category of modules over the (semi)ring `A` to the category of modules over
+    the `R`-algebra `A`. This is what is defined here, and is called `restrict_scalars` (though
+    perhaps a better name would be `enrich_scalars`).
+ 2. The functor from the category of modules over the `R`-algebra `A` to the category of modules
+    over the (semi)ring `R`. This is the true restriction of scalars but when one uses the setup
+    in the second example above, it is invisible functor available for free through typeclass
```

Comment:
```suggestion
    in the second example above, it is an invisible functor available for free through typeclass
```

Reply:
Somehow, despite my best efforts, I seem not to be able to communicate my point clearly.

I'll make one last attempt to find the right words but given that you are [not the first person](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/restrict_scalars.20vs.20enrich_scalars) who finds my words unclear, probably the thing to do is just to drop this remark entirely.

I claim the confusion stems from the fact that typeclass inference (in conjunction with mixins) makes a lot of forgetful functors invisible. To avoid this I'll try to make my point using informal mathematical language.

Thus let:

- `𝓜(S)` be the category of modules over a semiring `S`,
- `𝓡(A, R)` be the category of representations of an `R`-algebra `A` (over `R`),
- `F` be the functor that turns an algebra into a semiring (by forgetting the scalar action),
- `G : 𝓡(A, R) ⥤ 𝓜(R)` be the functor that forgets the `A` action.

Note there is a equivalence:

- `E : 𝓜(F(A)) ≌ 𝓡(A, R)`

The point I'm trying to make is that `restrict_scalars` is essentially `E` whereas true restriction of scalars is actually `G` (at least according to me).

Furthermore I claim that confusion about when to use `restrict_scalars` and when to use `is_scalar_tower` is the result of people invoking `E` because of the name of this type synonym, when what they really wanted was `G`.